### ISSUE TITLE
free(): invalid next size (fast)

### ISSUE THREAD

**Issue by def-** [ddnet/ddnet#8131]
Keeps crashing on RUS block2:
```
Core was generated by `./DDRace64-Server_sql -f servers/block2.cfg'.
Program terminated with signal SIGABRT, Aborted.
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0)
    at ./nptl/pthread_kill.c:44
44	./nptl/pthread_kill.c: No such file or directory.
[Current thread is 1 (Thread 0x7f00d9ffb6c0 (LWP 1519114))]
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0)
    at ./nptl/pthread_kill.c:44
#1  0x00007f00fc2a9e8f in __pthread_kill_internal (signo=6, threadid=<optimized out>)
    at ./nptl/pthread_kill.c:78
#2  0x00007f00fc25afb2 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007f00fc245472 in __GI_abort () at ./stdlib/abort.c:79
#4  0x00007f00fc29e430 in __libc_message (action=action@entry=do_abort,
    fmt=fmt@entry=0x7f00fc3b8459 "%s\n") at ../sysdeps/posix/libc_fatal.c:155
#5  0x00007f00fc2b37aa in malloc_printerr (
    str=str@entry=0x7f00fc3bb0c8 "free(): invalid next size (fast)") at ./malloc/malloc.c:5660
#6  0x00007f00fc2b561a in _int_free (av=0x7f00fc3f1c60 <main_arena>, p=0x55e1f550c7c0,
    have_lock=have_lock@entry=0) at ./malloc/malloc.c:4518
#7  0x00007f00fc2b7e8f in __GI___libc_free (mem=<optimized out>) at ./malloc/malloc.c:3385
#8  0x00007f00fc9ae956 in curl_slist_free_all () from /lib/x86_64-linux-gnu/libcurl.so.4
#9  0x000055e1f3e132eb in CHttpRequest::~CHttpRequest (this=0x55e1f55f4cc0, __in_chrg=<optimized out>)
    at /home/teeworlds/src/master/src/engine/shared/http.cpp:76
#10 CHttpRequest::~CHttpRequest (this=0x55e1f55f4cc0, __in_chrg=<optimized out>)
    at /home/teeworlds/src/master/src/engine/shared/http.cpp:78
#11 0x000055e1f3d8cd12 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (
    this=0x55e1f583a4c0) at /usr/include/c++/12/bits/shared_ptr_base.h:346
#12 std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x55e1f583a4c0)
    at /usr/include/c++/12/bits/shared_ptr_base.h:317
#13 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x55e1f5a41938,
    __in_chrg=<optimized out>) at /usr/include/c++/12/bits/shared_ptr_base.h:1071
#14 std::__shared_ptr<CHttpRequest, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x55e1f5a41930,
    __in_chrg=<optimized out>) at /usr/include/c++/12/bits/shared_ptr_base.h:1524
#15 std::shared_ptr<CHttpRequest>::~shared_ptr (this=0x55e1f5a41930, __in_chrg=<optimized out>)
    at /usr/include/c++/12/bits/shared_ptr.h:175
#16 CRegister::CProtocol::CJob::~CJob (this=0x55e1f5a418f0, __in_chrg=<optimized out>)
    at /home/teeworlds/src/master/src/engine/server/register.cpp:89
#17 std::_Destroy<CRegister::CProtocol::CJob> (__pointer=0x55e1f5a418f0)
    at /usr/include/c++/12/bits/stl_construct.h:151
#18 std::allocator_traits<std::allocator<void> >::destroy<CRegister::CProtocol::CJob> (
    __p=0x55e1f5a418f0) at /usr/include/c++/12/bits/alloc_traits.h:648
#19 std::_Sp_counted_ptr_inplace<CRegister::CProtocol::CJob, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (this=0x55e1f5a418e0) at /usr/include/c++/12/bits/shared_ptr_base.h:613
#20 0x000055e1f3e1ad10 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (
    this=0x55e1f5a418e0) at /usr/include/c++/12/bits/shared_ptr_base.h:346
#21 std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x55e1f5a418e0)
    at /usr/include/c++/12/bits/shared_ptr_base.h:317
#22 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=<synthetic pointer>,
    __in_chrg=<optimized out>) at /usr/include/c++/12/bits/shared_ptr_base.h:1071
#23 std::__shared_ptr<IJob, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=<synthetic pointer>,
    __in_chrg=<optimized out>) at /usr/include/c++/12/bits/shared_ptr_base.h:1524
#24 std::shared_ptr<IJob>::~shared_ptr (this=<synthetic pointer>, __in_chrg=<optimized out>)
    at /usr/include/c++/12/bits/shared_ptr.h:175
#25 CJobPool::RunLoop (this=0x55e1f5484c18) at /home/teeworlds/src/master/src/engine/shared/jobs.cpp:125
#26 0x000055e1f3e352c7 in thread_run (user=<optimized out>)
    at /home/teeworlds/src/master/src/base/system.cpp:794
#27 0x00007f00fc2a8134 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#28 0x00007f00fc3287dc in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
```

**Comment by def-** [comment]
Might be related to recent http changes

**Comment by Robyt3** [comment]
Possibly curl shuts down first (`curl_easy_cleanup` and `curl_global_cleanup`) which frees `CHttpRequest::m_pHeaders`, so the `CHttpRequest` destructor causes a double-free with its `curl_slist_free_all((curl_slist *)m_pHeaders)`.

**Comment by heinrich5991** [comment]
I don't understand your explanation for a double free. `curl_global_cleanup`/`curl_easy_cleanup` shouldn't clean up our member variable.

Calling `curl_slist_free_all` after curl was shut down is illegal though, so a problem anyway.

**Comment by Robyt3** [comment]
The way I understand/expect it, `curl_global_cleanup(pHandle)` will free some curl internal structure for the given handle and `curl_global_cleanup` will free all the curl internal structures, invalidating all handles. Hence, calling `curl_slist_free_all` after `curl_global_cleanup` seems wrong because the passed handle is not valid anymore. The assertion [`free(): invalid next size (fast)`](https://stackoverflow.com/a/4729416) indicates an incorrect pointer was passed to `free`, which in this case I assume is a double-free because we don't change the `m_pHeaders` itself to be invalid.
